File: desktop/macos/share-my-status-client/Core/StatusReporter.swift
//
//  StatusReporter.swift
//  share-my-status-client
//


import Foundation
import SwiftUI
import Combine

/// Main status reporter that coordinates all services
@MainActor
class StatusReporter: ObservableObject {
    // Published State
    @Published var isReporting = false
    @Published var lastError: Error?
    @Published var reportingStatus = "未启动"
    
    @Published var currentMusic: MusicSnapshot?
    @Published var currentSystem: SystemSnapshot?
    @Published var currentActivity: ActivitySnapshot?
    
    // Services
    private let mediaService: MediaRemoteService
    private let systemService: SystemMonitorService
    private let activityService: ActivityDetectorService
    private let networkService: NetworkService
    private let coverService: CoverService
    
    private let logger = AppLogger.reporter
    
    // Configuration
    private var configuration: AppConfiguration?
    
    // Store previous config state as values (not reference)
    private struct ConfigSnapshot {
        let musicReportingEnabled: Bool
        let systemReportingEnabled: Bool
        let activityReportingEnabled: Bool
        let systemPollingInterval: TimeInterval
        let activityPollingInterval: TimeInterval
        
        init(from config: AppConfiguration) {
            self.musicReportingEnabled = config.musicReportingEnabled
            self.systemReportingEnabled = config.systemReportingEnabled
            self.activityReportingEnabled = config.activityReportingEnabled
            self.systemPollingInterval = config.systemPollingInterval
            self.activityPollingInterval = config.activityPollingInterval
        }
    }
    private var previousConfigSnapshot: ConfigSnapshot?
    
    // Report Timers (for polling services)
    private var systemReportTimer: Timer?
    private var activityReportTimer: Timer?
    
    // Initialization
    init() {
        self.mediaService = MediaRemoteService()
        self.systemService = SystemMonitorService()
        self.activityService = ActivityDetectorService()
        self.networkService = NetworkService()
        self.coverService = CoverService()
        
        logger.info("StatusReporter initialized")
    }
    
    deinit {
        // Clean up timers
        systemReportTimer?.invalidate()
        activityReportTimer?.invalidate()
        
        // Note: Cannot call async stopReporting() from deinit
        // Services will clean up themselves via their own deinit
    }
    
    // Track if configuration update is in progress
    private var isUpdatingConfiguration = false
    
    // Cache for deduplication
    private var lastReportedActivityLabel: String?

    // Configuration Update
    func updateConfiguration(_ config: AppConfiguration) {
        // Prevent concurrent configuration updates
        guard !isUpdatingConfiguration else {
            logger.warning("Configuration update already in progress, skipping")
            return
        }
        
        isUpdatingConfiguration = true
        
        // Capture previous config snapshot (values, not reference)
        let previousSnapshot = previousConfigSnapshot
        
        // Update configuration reference
        self.configuration = config
        
        // Save new snapshot for next update
        previousConfigSnapshot = ConfigSnapshot(from: config)
        
        Task {
            defer { 
                Task { @MainActor in
                    self.isUpdatingConfiguration = false
                }
            }
            
            logger.info("Updating configuration...")
            
            // Update all services
            await mediaService.updateWhitelist(config.musicAppWhitelist)
            await activityService.updateActivityGroups(config.activityGroups)
            await networkService.updateConfiguration(
                endpointURL: config.endpointURL,
                secretKey: config.secretKey
            )
            await coverService.updateConfiguration(
                baseURL: config.endpointURL,
                secretKey: config.secretKey
            )
            
            // Update polling intervals
            await systemService.updatePollingInterval(config.systemPollingInterval)
            await activityService.updatePollingInterval(config.activityPollingInterval)
            
            // Handle reporting state changes
            if !config.isReportingEnabled && isReporting {
                // User disabled reporting entirely
                logger.info("Reporting disabled, stopping all services")
                stopReporting()
            } else if config.isReportingEnabled && !isReporting {
                // User enabled reporting
                logger.info("Reporting enabled, starting services")
                startReporting()
            } else if config.isReportingEnabled && isReporting {
                // Reporting is active, check individual service toggles
                logger.info("Reporting active, checking service toggles")
                await handleServiceToggles(previous: previousSnapshot, current: config)
            }
            
            logger.info("Configuration update completed")
        }
    }
    
    // Handle Individual Service Toggles
    private func handleServiceToggles(previous: ConfigSnapshot?, current: AppConfiguration) async {
        guard let prev = previous else {
            // No previous config - this is initial startup
            // Services should already be started by the initial startReporting() call
            logger.info("No previous configuration, services should already be running")
            return
        }
        
        logger.info("Checking service toggles...")
        logger.info("Music: prev=\(prev.musicReportingEnabled), current=\(current.musicReportingEnabled)")
        logger.info("System: prev=\(prev.systemReportingEnabled), current=\(current.systemReportingEnabled)")
        logger.info("Activity: prev=\(prev.activityReportingEnabled), current=\(current.activityReportingEnabled)")
        
        // Check music service toggle
        if prev.musicReportingEnabled != current.musicReportingEnabled {
            if current.musicReportingEnabled {
                logger.info("Starting music service (user enabled)...")
                await startMusicService()
            } else {
                logger.info("Stopping music service (user disabled)...")
                await stopMusicService()
            }
        } else {
            logger.debug("Music service toggle unchanged")
        }
        
        // Check system service toggle
        if prev.systemReportingEnabled != current.systemReportingEnabled {
            if current.systemReportingEnabled {
                logger.info("Starting system service (user enabled)...")
                await startSystemService()
            } else {
                logger.info("Stopping system service (user disabled)...")
                await stopSystemService()
            }
        }
        
        // Check activity service toggle
        if prev.activityReportingEnabled != current.activityReportingEnabled {
            if current.activityReportingEnabled {
                logger.info("Starting activity service (user enabled)...")
                await startActivityService()
            } else {
                logger.info("Stopping activity service (user disabled)...")
                await stopActivityService()
            }
        }
        
        // Check if polling intervals changed (restart if running)
        if prev.systemPollingInterval != current.systemPollingInterval && current.systemReportingEnabled {
            logger.info("System polling interval changed, restarting...")
            await stopSystemService()
            await startSystemService()
        }
        
        if prev.activityPollingInterval != current.activityPollingInterval && current.activityReportingEnabled {
            logger.info("Activity polling interval changed, restarting...")
            await stopActivityService()
            await startActivityService()
        }
        
        logger.info("Service toggles handling completed")
        await updateReportingStatus()
    }
    
    // Start Reporting
    func startReporting() {
        guard let config = configuration else {
            logger.error("No configuration available")
            reportingStatus = "未配置"
            return
        }
        
        guard config.isValidConfiguration() else {
            logger.error("Invalid configuration")
            reportingStatus = "配置不完整"
            return
        }
        
        logger.info("Starting status reporting...")
        isReporting = true
        lastError = nil
        
        Task {
            // Start each enabled service using dedicated methods
            if config.musicReportingEnabled {
                await startMusicService()
            } else {
                logger.info("Music reporting disabled in config")
            }
            
            if config.systemReportingEnabled {
                await startSystemService()
            } else {
                logger.info("System reporting disabled in config")
            }
            
            if config.activityReportingEnabled {
                await startActivityService()
            } else {
                logger.info("Activity reporting disabled in config")
            }
            
            await updateReportingStatus()
        }
    }
    
    // Stop Reporting
    func stopReporting() {
        logger.info("Stopping status reporting...")
        isReporting = false
        
        Task {
            // Stop all services using dedicated methods
            await stopMusicService()
            await stopSystemService()
            await stopActivityService()
            
            await updateReportingStatus()
        }
    }
    
    // Individual Service Control
    
    /// Start music service only
    private func startMusicService() async {
        guard let config = configuration, config.musicReportingEnabled else { 
            logger.info("Music reporting disabled, skipping start")
            return 
        }
        
        // Check if already running
        let isAlreadyStreaming = await mediaService.getIsStreaming()
        if isAlreadyStreaming {
            logger.warning("Music service already streaming, skipping start")
            return
        }
        
        logger.info("Starting music service...")
        do {
            // Register callback for music changes
            await mediaService.registerCallback { [weak self] music in
                Task { @MainActor [weak self] in
                    guard let self = self else { return }
                    
                    // Check if this is only an artwork update (same song)
                    let isArtworkOnlyUpdate = music != nil && 
                                             self.currentMusic != nil &&
                                             music!.title == self.currentMusic!.title &&
                                             music!.artist == self.currentMusic!.artist
                    
                    if let music = music {
                        if isArtworkOnlyUpdate {
                            self.logger.info("Artwork update for: \(music.artist) - \(music.title)")
                        } else {
                            self.logger.info("Music change event: \(music.artist) - \(music.title)")
                        }
                    } else {
                        self.logger.info("Music change event: no music playing")
                    }
                    
                    // Always update current music state
                    self.currentMusic = music
                    
                    // Only report if it's not just an artwork update
                    if !isArtworkOnlyUpdate {
                        await self.reportMusicChange(music)
                    } else {
                        self.logger.info("Skipping report for artwork-only update (will be included in next report)")
                    }
                }
            }
            
            // Start streaming
            try await mediaService.start()
            logger.info("Music service started successfully")
        } catch {
            logger.error("Failed to start music service: \(error)")
        }
    }
    
    /// Stop music service only
    private func stopMusicService() async {
        logger.info("Stopping music service...")
        
        // Check if music service is actually running
        let isStreaming = await mediaService.getIsStreaming()
        logger.info("Music service streaming status before stop: \(isStreaming)")
        
        await mediaService.stop()
        
        // Verify it stopped
        let isStillStreaming = await mediaService.getIsStreaming()
        logger.info("Music service streaming status after stop: \(isStillStreaming)")
        
        currentMusic = nil
        logger.info("Music service stopped")
    }
    
    /// Start system monitoring service only
    private func startSystemService() async {
        guard let config = configuration, config.systemReportingEnabled else { return }
        
        logger.info("Starting system service...")
        do {
            await systemService.updatePollingInterval(config.systemPollingInterval)
            try await systemService.start()
            await setupSystemReportTimer(interval: config.systemPollingInterval)
            logger.info("System service started")
        } catch {
            logger.error("Failed to start system service: \(error)")
        }
    }
    
    /// Stop system monitoring service only
    private func stopSystemService() async {
        logger.info("Stopping system service...")
        await systemService.stop()
        systemReportTimer?.invalidate()
        systemReportTimer = nil
        currentSystem = nil
    }
    
    /// Start activity detection service only
    private func startActivityService() async {
        guard let config = configuration, config.activityReportingEnabled else { return }
        
        logger.info("Starting activity service...")
        do {
            await activityService.updatePollingInterval(config.activityPollingInterval)
            try await activityService.start()
            await setupActivityReportTimer(interval: config.activityPollingInterval)
            logger.info("Activity service started")
        } catch {
            logger.error("Failed to start activity service: \(error)")
        }
    }
    
    /// Stop activity detection service only
    private func stopActivityService() async {
        logger.info("Stopping activity service...")
        await activityService.stop()
        activityReportTimer?.invalidate()
        activityReportTimer = nil
        currentActivity = nil
        
        // Clear cached label when stopping service
        lastReportedActivityLabel = nil
    }
    
    // Report Timers
    
    /// Setup system monitoring report timer (polling-based)
    private func setupSystemReportTimer(interval: TimeInterval) async {
        systemReportTimer?.invalidate()
        
        systemReportTimer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in
                await self?.reportSystemStatus()
            }
        }
        
        logger.info("System report timer set to \(interval) seconds")
    }
    
    /// Setup activity detection report timer (polling-based)
    private func setupActivityReportTimer(interval: TimeInterval) async {
        activityReportTimer?.invalidate()
        
        activityReportTimer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in
                await self?.reportActivityStatus()
            }
        }
        
        logger.info("Activity report timer set to \(interval) seconds")
    }
    
    // Report Methods
    
    /// Report music change immediately (event-driven)
    private func reportMusicChange(_ music: MusicSnapshot?) async {
        guard let config = configuration, config.isReportingEnabled else {
            return
        }
        
        logger.debug("Reporting music change...")
        
        // If no music is playing and we don't have previous music, skip reporting
        // This avoids reporting empty state on app startup
        if music == nil && currentMusic == nil {
            logger.info("No music playing and no previous music, skipping report")
            return
        }
        
        var musicInfo: MusicInfo? = nil
        
        // Use currentMusic which may have been updated with artwork via diff updates
        if let music = currentMusic {
            logger.info("Music changed: \(music.artist) - \(music.title), playing: \(music.isPlaying)")
            var coverHash: String? = nil
            
            // Upload cover if we have artwork data
            if let artworkData = music.artworkData {
                logger.info("Uploading cover (size: \(artworkData.count) bytes)")
                do {
                    coverHash = try await coverService.checkAndUploadCover(artworkData: artworkData)
                    logger.info("Cover uploaded with hash: \(coverHash ?? "nil")")
                } catch {
                    logger.error("Failed to upload cover: \(error)")
                }
            } else {
                logger.info("No artwork data available for current music")
            }
            
            musicInfo = music.toMusicInfo(coverHash: coverHash)
        } else {
            logger.info("Music stopped (was playing: \(music?.title ?? "unknown"))")
        }
        
        // Create report event with only music info
        let event = ReportEvent(
            system: nil,
            music: musicInfo,
            activity: nil
        )
        
        await sendReport(event: event, source: "music")
    }
    
    /// Report system status periodically (polling-based)
    private func reportSystemStatus() async {
        guard let config = configuration, config.isReportingEnabled, config.systemReportingEnabled else {
            return
        }
        
        logger.debug("Reporting system status...")
        
        // Get current system snapshot
        currentSystem = await systemService.getCurrentSnapshot()
        guard let systemInfo = currentSystem?.toSystemInfo() else {
            logger.info("No system data to report")
            return
        }
        
        logger.info("System status collected")
        
        // Create report event with only system info
        let event = ReportEvent(
            system: systemInfo,
            music: nil,
            activity: nil
        )
        
        await sendReport(event: event, source: "system")
    }
    
    /// Report activity status periodically (polling-based)
    private func reportActivityStatus() async {
        guard let config = configuration, config.isReportingEnabled, config.activityReportingEnabled else {
            return
        }
        
        logger.debug("Reporting activity status...")
        
        // Get current activity
        currentActivity = await activityService.getCurrentActivity()
        guard let activityInfo = currentActivity?.toActivityInfo() else {
            logger.info("No activity data to report")
            return
        }
        
        // Check if label has changed (deduplication)
        if let lastLabel = lastReportedActivityLabel, lastLabel == activityInfo.label {
            logger.debug("Activity label unchanged (\(activityInfo.label)), skipping report")
            return
        }
        
        logger.info("Activity status collected: \(activityInfo.label)")
        
        // Update last reported label
        lastReportedActivityLabel = activityInfo.label
        
        // Create report event with only activity info
        let event = ReportEvent(
            system: nil,
            music: nil,
            activity: activityInfo
        )
        
        await sendReport(event: event, source: "activity")
    }
    
    /// Send report to server
    private func sendReport(event: ReportEvent, source: String) async {
        let request = BatchReportRequest(events: [event])
        
        do {
            let response = try await networkService.reportStatus(request)
            lastError = nil
            
            logger.info("[\(source)] Report sent successfully: accepted=\(response.accepted ?? 0)")
            
            await updateReportingStatus()
        } catch {
            logger.error("[\(source)] Failed to send report: \(error)")
            lastError = error
            await updateReportingStatus()
        }
    }
    
    // Update Status
    private func updateReportingStatus() async {
        guard let config = configuration else {
            reportingStatus = "未配置"
            return
        }
        
        if !config.isReportingEnabled {
            reportingStatus = "已禁用"
            return
        }
        
        if !isReporting {
            reportingStatus = "未启动"
            return
        }
        
        let connected = await networkService.getConnectionStatus()
        if !connected {
            reportingStatus = "网络未连接"
            return
        }
        
        if let error = lastError {
            reportingStatus = "错误: \(error.localizedDescription)"
            return
        }
        
        var activeModules: [String] = []
        
        if config.musicReportingEnabled, await mediaService.isActive() {
            activeModules.append("音乐")
        }
        
        if config.systemReportingEnabled, await systemService.isActive() {
            activeModules.append("系统")
        }
        
        if config.activityReportingEnabled, await activityService.isActive() {
            activeModules.append("活动")
        }
        
        if activeModules.isEmpty {
            reportingStatus = "无活动模块"
        } else {
            reportingStatus = "正在上报: \(activeModules.joined(separator: ", "))"
        }
    }
    
    // Status Summary
    func getStatusSummary() -> String {
        var summary: [String] = []
        
        if let music = currentMusic {
            summary.append("🎵 \(music.artist) - \(music.title)")
        }
        
        if let system = currentSystem {
            var systemInfo: [String] = []
            if let battery = system.batteryPercentage {
                let icon = system.isCharging == true ? "🔌" : "🔋"
                systemInfo.append("\(icon) \(battery)%")
            }
            if let cpu = system.cpuPercentage {
                systemInfo.append("💻 CPU \(cpu)%")
            }
            if let memory = system.memoryPercentage {
                systemInfo.append("🧠 内存 \(memory)%")
            }
            if !systemInfo.isEmpty {
                summary.append(systemInfo.joined(separator: " "))
            }
        }
        
        if let activity = currentActivity {
            let icon = activity.isIdle ? "😴" : "👤"
            summary.append("\(icon) \(activity.activityTag): \(activity.activeApplication)")
        }
        
        return summary.isEmpty ? "无状态数据" : summary.joined(separator: "\n")
    }
    
    // Network Statistics
    func getNetworkStatistics() async -> (lastReportTime: Date?, reportCount: Int, isConnected: Bool) {
        return await networkService.getStatistics()
    }
}

File: desktop/windows/ShareMyStatusClient/Services/StatusReporter.cs
using ShareMyStatusClient.Models.Api;
using ShareMyStatusClient.Models.Domain;
using ShareMyStatusClient.Models.Settings;
using ShareMyStatusClient.Utilities;

namespace ShareMyStatusClient.Services;

/// <summary>
/// Coordinates the collection services and reports to the backend. Mirrors the macOS
/// StatusReporter: system + activity are polled on dedicated loops, music is event-driven.
/// </summary>
public sealed class StatusReporter : IDisposable
{
    private readonly SystemMonitorService _system = new();
    private readonly ActivityDetectorService _activity = new();
    private readonly MediaSessionService _media = new();
    private readonly NetworkService _network = new();
    private readonly CoverService _cover = new();
    private readonly AppLogger _logger = AppLogger.Reporter;

    private readonly object _stateGate = new();
    // Serializes Start/Stop/Restart so configuration changes can't interleave loops.
    private readonly SemaphoreSlim _lifecycleLock = new(1, 1);
    // Serializes event-driven music handling so concurrent GSMTC callbacks can't
    // reorder or double-report.
    private readonly SemaphoreSlim _musicLock = new(1, 1);

    private AppConfiguration _config = new();

    private CancellationTokenSource? _systemCts;
    private CancellationTokenSource? _activityCts;
    private Task? _systemTask;
    private Task? _activityTask;
    private string? _lastReportedActivityLabel;

    // Backing fields for cross-thread-observed state (guarded by _stateGate).
    private bool _isReporting;
    private string _reportingStatus = "未启动";
    private Exception? _lastError;

    private MusicSnapshot? _currentMusic;
    private SystemSnapshot? _currentSystem;
    private ActivitySnapshot? _currentActivity;
    private ClientResources? _clientResources;

    public bool IsReporting
    {
        get { lock (_stateGate) { return _isReporting; } }
        private set { lock (_stateGate) { _isReporting = value; } }
    }

    public string ReportingStatus
    {
        get { lock (_stateGate) { return _reportingStatus; } }
    }

    public Exception? LastError
    {
        get { lock (_stateGate) { return _lastError; } }
        private set { lock (_stateGate) { _lastError = value; } }
    }

    public MusicSnapshot? CurrentMusic
    {
        get { lock (_stateGate) { return _currentMusic; } }
    }

    public SystemSnapshot? CurrentSystem
    {
        get { lock (_stateGate) { return _currentSystem; } }
    }

    public ActivitySnapshot? CurrentActivity
    {
        get { lock (_stateGate) { return _currentActivity; } }
    }

    /// <summary>Server-provided resource links (user status page / signature DIY), if fetched.</summary>
    public ClientResources? ClientResources
    {
        get { lock (_stateGate) { return _clientResources; } }
        private set { lock (_stateGate) { _clientResources = value; } }
    }

    /// <summary>Raised whenever observable state changes (status, current snapshots).</summary>
    public event EventHandler? Changed;

    // ---- Configuration ----

    public void ApplyConfiguration(AppConfiguration config)
    {
        var clone = config.Clone();
        lock (_stateGate) { _config = clone; }

        _network.UpdateConfiguration(clone.EndpointUrl, clone.SecretKey);
        _cover.UpdateConfiguration(clone.EndpointUrl, clone.SecretKey);
        _media.UpdateWhitelist(clone.MusicAppWhitelist);

        _logger.Info("Configuration applied");

        // Refresh server resource links in the background (best-effort).
        _ = RefreshClientResourcesAsync();

        if (IsReporting)
        {
            // Restart loops to pick up new toggles/intervals (serialized + race-free).
            _ = RestartAsync();
        }
    }

    /// <summary>Validate the given endpoint + secret against the server (used by the settings dialog).</summary>
    public Task<(bool Ok, string Message)> TestConnectionAsync(string endpointUrl, string secretKey) =>
        _network.TestConnectionAsync(endpointUrl, secretKey, CancellationToken.None);

    public async Task RefreshClientResourcesAsync()
    {
        try
        {
            var resources = await _network.FetchClientResourcesAsync(CancellationToken.None).ConfigureAwait(false);
            if (resources != null)
            {
                ClientResources = resources;
                RaiseChanged();
            }
        }
        catch (Exception ex)
        {
            _logger.Debug($"Client resources fetch failed: {ex.Message}");
        }
    }

    // ---- Lifecycle (serialized through _lifecycleLock) ----

    public async Task StartAsync()
    {
        await _lifecycleLock.WaitAsync().ConfigureAwait(false);
        try { await StartCoreAsync().ConfigureAwait(false); }
        finally { _lifecycleLock.Release(); }
    }

    public async Task StopAsync()
    {
        await _lifecycleLock.WaitAsync().ConfigureAwait(false);
        try { await StopCoreAsync().ConfigureAwait(false); }
        finally { _lifecycleLock.Release(); }
    }

    private async Task RestartAsync()
    {
        await _lifecycleLock.WaitAsync().ConfigureAwait(false);
        try
        {
            await StopCoreAsync().ConfigureAwait(false);
            await StartCoreAsync().ConfigureAwait(false);
        }
        finally { _lifecycleLock.Release(); }
    }

    private async Task StartCoreAsync()
    {
        AppConfiguration cfg;
        lock (_stateGate) { cfg = _config; }

        if (!cfg.IsValid())
        {
            _logger.Error("Invalid configuration; cannot start");
            SetStatus("配置不完整");
            return;
        }

        _logger.Info("Starting status reporting...");
        IsReporting = true;
        LastError = null;

        if (cfg.MusicReportingEnabled)
            await _media.StartAsync(OnMusicChanged).ConfigureAwait(false);

        if (cfg.SystemReportingEnabled)
            StartSystemLoop(cfg);

        if (cfg.ActivityReportingEnabled)
            StartActivityLoop(cfg);

        UpdateStatus();
    }

    private async Task StopCoreAsync()
    {
        _logger.Info("Stopping status reporting...");
        IsReporting = false;

        _systemCts?.Cancel();
        _activityCts?.Cancel();

        // Wait for the loops to actually exit before disposing their token sources.
        var pending = new[] { _systemTask, _activityTask }.Where(t => t != null).Cast<Task>().ToArray();
        if (pending.Length > 0)
        {
            try { await Task.WhenAll(pending).ConfigureAwait(false); }
            catch (OperationCanceledException) { /* expected */ }
            catch (Exception ex) { _logger.Debug($"Loop drain error: {ex.Message}"); }
        }

        _systemCts?.Dispose();
        _systemCts = null;
        _systemTask = null;
        _activityCts?.Dispose();
        _activityCts = null;
        _activityTask = null;

        await _media.StopAsync().ConfigureAwait(false);

        lock (_stateGate)
        {
            _currentMusic = null;
            _currentSystem = null;
            _currentActivity = null;
            _lastReportedActivityLabel = null;
        }

        UpdateStatus();
    }

    // ---- System loop ----

    private void StartSystemLoop(AppConfiguration cfg)
    {
        _system.Reset();
        var cts = new CancellationTokenSource();
        _systemCts = cts;
        var interval = TimeSpan.FromSeconds(cfg.SystemPollingInterval);

        _systemTask = Task.Run(async () =>
        {
            while (!cts.IsCancellationRequested)
            {
                try
                {
                    var snapshot = _system.Collect();
                    lock (_stateGate) { _currentSystem = snapshot; }
                    RaiseChanged();
                    await SendReportAsync(ReportEvent.ForSystem(snapshot.ToSystemInfo()), "system", cts.Token)
                        .ConfigureAwait(false);
                }
                catch (Exception ex)
                {
                    _logger.Error("System loop error", ex);
                }

                try { await Task.Delay(interval, cts.Token).ConfigureAwait(false); }
                catch (TaskCanceledException) { break; }
            }
        }, cts.Token);
    }

    // ---- Activity loop ----

    private void StartActivityLoop(AppConfiguration cfg)
    {
        var cts = new CancellationTokenSource();
        _activityCts = cts;
        var interval = TimeSpan.FromSeconds(cfg.ActivityPollingInterval);

        _activityTask = Task.Run(async () =>
        {
            while (!cts.IsCancellationRequested)
            {
                try
                {
                    List<ActivityGroup> groups;
                    lock (_stateGate) { groups = _config.ActivityGroups; }

                    var snapshot = _activity.Collect(groups);
                    if (snapshot != null)
                    {
                        string? lastLabel;
                        lock (_stateGate)
                        {
                            _currentActivity = snapshot;
                            lastLabel = _lastReportedActivityLabel;
                        }
                        RaiseChanged();

                        if (lastLabel != snapshot.ActivityTag)
                        {
                            // Only mark the label as reported after a successful send, so a
                            // failed report doesn't permanently suppress re-reporting it.
                            var ok = await SendReportAsync(ReportEvent.ForActivity(snapshot.ToActivityInfo()), "activity", cts.Token)
                                .ConfigureAwait(false);
                            if (ok)
                                lock (_stateGate) { _lastReportedActivityLabel = snapshot.ActivityTag; }
                        }
                    }
                }
                catch (Exception ex)
                {
                    _logger.Error("Activity loop error", ex);
                }

                try { await Task.Delay(interval, cts.Token).ConfigureAwait(false); }
                catch (TaskCanceledException) { break; }
            }
        }, cts.Token);
    }

    // ---- Music (event-driven) ----

    private void OnMusicChanged(MusicSnapshot? music)
    {
        _ = HandleMusicAsync(music);
    }

    private async Task HandleMusicAsync(MusicSnapshot? music)
    {
        if (!IsReporting)
            return; // ignore late callbacks after stop

        // Serialize concurrent GSMTC callbacks so reports can't reorder or duplicate.
        await _musicLock.WaitAsync().ConfigureAwait(false);
        try
        {
            if (!IsReporting)
                return;

            bool artworkOnly;
            lock (_stateGate)
            {
                var current = _currentMusic;
                var sameSong = music != null && current != null
                    && music.Title == current.Title
                    && music.Artist == current.Artist;
                // GSMTC often delivers the thumbnail in a later event than the metadata.
                // If the cover only just arrived for the same song, still report it once so
                // the cover isn't lost (the macOS client coordinates this via waitingForArtwork).
                var artworkNewlyArrived = sameSong
                    && (current!.ArtworkData == null || current.ArtworkData.Length == 0)
                    && music!.ArtworkData is { Length: > 0 };
                artworkOnly = sameSong && !artworkNewlyArrived;
                _currentMusic = music;
            }

            RaiseChanged();

            if (!artworkOnly)
                await ReportMusicChangeAsync().ConfigureAwait(false);
        }
        finally
        {
            _musicLock.Release();
        }
    }

    private async Task ReportMusicChangeAsync()
    {
        if (!IsReporting)
            return;

        MusicSnapshot? current;
        lock (_stateGate) { current = _currentMusic; }

        // No current music (e.g. playback stopped) => nothing to report, matching macOS.
        if (current == null)
            return;

        string? coverHash = null;
        if (current.ArtworkData is { Length: > 0 } artwork)
        {
            try
            {
                coverHash = await _cover.CheckAndUploadCoverAsync(artwork, CancellationToken.None).ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                _logger.Error("Failed to upload cover", ex);
            }
        }

        var info = current.ToMusicInfo(coverHash);
        await SendReportAsync(ReportEvent.ForMusic(info), "music", CancellationToken.None).ConfigureAwait(false);
    }

    // ---- Reporting ----

    private async Task<bool> SendReportAsync(ReportEvent ev, string source, CancellationToken ct)
    {
        var request = new BatchReportRequest();
        request.Events.Add(ev);

        var ok = false;
        try
        {
            var response = await _network.ReportStatusAsync(request, ct).ConfigureAwait(false);
            LastError = null;
            ok = true;
            _logger.Info($"[{source}] report sent: accepted={response.Accepted ?? 0}");
        }
        catch (OperationCanceledException)
        {
            return false; // shutting down; don't touch status
        }
        catch (Exception ex)
        {
            LastError = ex;
            _logger.Error($"[{source}] report failed", ex);
        }

        UpdateStatus();
        return ok;
    }

    // ---- Status ----

    private void UpdateStatus()
    {
        string status;
        AppConfiguration cfg;
        bool reporting;
        Exception? error;
        lock (_stateGate)
        {
            cfg = _config;
            reporting = _isReporting;
            error = _lastError;
        }

        if (!reporting)
        {
            status = "未启动";
        }
        else if (!NetworkService.IsConnected)
        {
            status = "网络未连接";
        }
        else if (error != null)
        {
            status = $"错误: {error.Message}";
        }
        else
        {
            var modules = new List<string>();
            // For music, reflect whether the media listener actually started (matches macOS).
            if (cfg.MusicReportingEnabled && _media.IsActive) modules.Add("音乐");
            if (cfg.SystemReportingEnabled) modules.Add("系统");
            if (cfg.ActivityReportingEnabled) modules.Add("活动");
            status = modules.Count == 0 ? "无活动模块" : "正在上报: " + string.Join(", ", modules);
        }

        SetStatus(status);
    }

    private void SetStatus(string status)
    {
        lock (_stateGate) { _reportingStatus = status; }
        RaiseChanged();
    }

    private void RaiseChanged() => Changed?.Invoke(this, EventArgs.Empty);

    /// <summary>Human-readable one-shot summary for the tray tooltip.</summary>
    public string GetStatusSummary()
    {
        var parts = new List<string>();

        MusicSnapshot? music;
        SystemSnapshot? system;
        ActivitySnapshot? activity;
        lock (_stateGate)
        {
            music = _currentMusic;
            system = _currentSystem;
            activity = _currentActivity;
        }

        if (music != null)
            parts.Add($"🎵 {music.Artist} - {music.Title}");

        if (system != null)
        {
            var sys = new List<string>();
            if (system.BatteryPercentage is { } b)
                sys.Add($"{(system.IsCharging == true ? "🔌" : "🔋")} {b}%");
            if (system.CpuPercentage is { } c)
                sys.Add($"💻 CPU {c}%");
            if (system.MemoryPercentage is { } m)
                sys.Add($"🧠 内存 {m}%");
            if (sys.Count > 0)
                parts.Add(string.Join(" ", sys));
        }

        if (activity != null)
            parts.Add($"{(activity.IsIdle ? "😴" : "👤")} {activity.ActivityTag}: {activity.ActiveApplication}");

        return parts.Count == 0 ? "无状态数据" : string.Join("\n", parts);
    }

    /// <summary>Releases owned services and synchronization primitives.
    /// Caller must have stopped reporting first (App does this on exit).</summary>
    public void Dispose()
    {
        _network.Dispose();
        _cover.Dispose();
        _media.Dispose();
        _lifecycleLock.Dispose();
        _musicLock.Dispose();
    }
}
